### PR TITLE
Add feature id to annotations; prepare showcase of callout annotations

### PR DESCRIPTION
**Pull request by kiryldz**
### Summary of changes

Expose unique feature identifier for all annotations created by annotation plugin. This feature id is used when preparing a `Feature` that backs up annotation. This feature id could now be used to connect annotation with view annotation.

Prepare an example of annotation usage (drawing a marker) and associated view annotation.

https://user-images.githubusercontent.com/15800566/146143453-be75920c-9e70-41ba-8e88-873ecb69acfc.mp4

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
    
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [x] Add example if relevant.
 - [x] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [x] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog>Expose unique annotation feature identifier that could be used to link view annotation to any annotation.</changelog>`.
 - [ ] If this PR is a `v10.[version]` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.[version]` release branch.

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

### CODE REVIEW

File: sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/Annotation.kt [hunk 1]
```diff
@@ -87,5 +92,10 @@ abstract class Annotation<T : Geometry>(
      * Minimum latitude value in Mercator projection.
      */
     const val MIN_MERCATOR_LATITUDE = -85.05112877980659
+
+    /**
+     * Unique feature identifier incremental counter.
+     */
+    private var FEATURE_IDENTIFIER = 5234
```
Comment (alexander-kulikovskii) [inline sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/Annotation.kt:99]: should it be in `companion` block? I guess you want to get unique id for each instance of objects

Comment (kiryldz) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/Annotation.kt:99]: code in the activity is not ideal for now, I'll revisit it and introduce smaller functions a bit later.

File: sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/Annotation.kt [hunk 2]
```diff
@@ -71,6 +71,11 @@ abstract class Annotation<T : Geometry>(
    */
   var isSelected: Boolean = false
 
+  /**
+   * Unique feature identifier.
+   */
+  val featureIdentifier: String = "annotation_feature_${(FEATURE_IDENTIFIER++)}"
```
Comment (Chaoba) [inline sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/Annotation.kt:77]: Can we reuse `id`?

Comment (kiryldz) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/Annotation.kt:77]: hmm, nice, that way we technically don't even need new API if I understand everything correctly

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt [hunk 3]
```diff
@@ -453,7 +453,7 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
   private fun convertAnnotationsToFeatures(annotations: Collection<T>): List<Feature> =
     annotations.map {
       it.setUsedDataDrivenProperties()
-      Feature.fromGeometry(it.geometry, it.getJsonObjectCopy())
+      Feature.fromGeometry(it.geometry, it.getJsonObjectCopy(), it.featureIdentifier)
```
Comment (pengdev) [inline plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:456]: @Chaoba do we use generate ID for annotation plugin?

Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:456]: We can update Annotation to let the id starts from a big number instead of 0

File: sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/Annotation.kt (hunk 1 above)
Comment (kiryldz) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/Annotation.kt]: revisited

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt (hunk 3 above)
Comment (kiryldz) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:456]: that's still imho pretty unreliable, PTAL at my update here

Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:456]: The new solution LGTM.

File: sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/Annotation.kt [hunk 4]
```diff
@@ -87,5 +92,7 @@ abstract class Annotation<T : Geometry>(
      * Minimum latitude value in Mercator projection.
      */
     const val MIN_MERCATOR_LATITUDE = -85.05112877980659
+
+    private const val FEATURE_IDENTIFIER_PREFIX = "mapbox_annotation_"
```
Comment (ank27) [inline sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/Annotation.kt:96]: nit: we can add documentation here too, for consistency.

Comment (kiryldz) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/annotation/Annotation.kt:96]: imho naming is self-descriptive and it's private 😉

**Review by ank27 — APPROVED** [review]

**Comment by macdrevx** [comment]
The iOS SDK has used the annotation id as the feature ID since v10.0.0: https://github.com/mapbox/mapbox-maps-ios/blob/v10.0.0/Sources/MapboxMaps/Annotations/Generated/PointAnnotation.swift#L25

It defaults the IDs to a UUID string, but developers can provide any string they like: https://github.com/mapbox/mapbox-maps-ios/blob/main/Sources/MapboxMaps/Annotations/Generated/PointAnnotation.swift#L45

Some alignment questions:
* Is it necessary to prefix the ID? If not, we could be aligned without any changes to iOS.
* If we choose not to prefix, could we eliminate the new featureIdentifier property and just document that `id` is used as the underlying feature ID?

**Comment by kiryldz** [comment]
> The iOS SDK has used the annotation id as the feature ID since v10.0.0: https://github.com/mapbox/mapbox-maps-ios/blob/v10.0.0/Sources/MapboxMaps/Annotations/Generated/PointAnnotation.swift#L25
> 
> It defaults the IDs to a UUID string, but developers can provide any string they like: https://github.com/mapbox/mapbox-maps-ios/blob/main/Sources/MapboxMaps/Annotations/Generated/PointAnnotation.swift#L45
> 
> Some alignment questions:
> 
> * Is it necessary to prefix the ID? If not, we could be aligned without any changes to iOS.
> * If we choose not to prefix, could we eliminate the new featureIdentifier property and just document that `id` is used as the underlying feature ID?
> 
> Overall, I like the concept.

@macdrevx thanks for looking into it!
I was also thinking of actually using our `Annotation.id` exposed already but my main concern here is that we generate it not as UUID but simply as 0, 1, 2 etc. which (at least for me) looks a bit unreliable to use such an easy id when preparing a `Feature` (using static method `Feature.fromGeometry(geometry, properties, id)`).

At the same time we can't re-use `Annotation.id` making generating UUID more complex from our end as it is `Long` and not `String`. That's actually why current solution is used: concat unique prefix + incremental annotation id we generate.

To sum up, at least from my understanding, perhaps there's no need to even add anything on iOS then. It would be nice to prepare similar example to show users how to use point annotation with an icon with view annotation but no need to align on `Annotation.featureIdentifier` as it seems we're not perfectly aligned anyway (e.g. our `id` is `Long`, there's no way to allow user provide their own feature id as we don't even assign feature an id before this PR).

**Comment by macdrevx** [comment]
Sounds good to me. Could you open a ticket for iOS to add the example? cc: @endanke is that something you could do?

**Review by Chaoba — APPROVED** [review]

**Comment by endanke** [comment]
@macdrevx @kiryldz sure! I'll open a ticket and assign myself. Although I don't think I can work on it before my holidays, but I'll pick it up in January.

**Review by yunikkk — APPROVED** [review]
